Tasks/ContainerBuildV0/src/dockerbuild.ts

```diff
@@ -0,0 +1,58 @@
+"use strict";
+
+import * as fs from "fs";
+import * as path from "path";
+import * as tl from "azure-pipelines-task-lib/task";
+import ContainerConnection from "docker-common-v2/containerconnection";
+import * as dockerCommandUtils from "docker-common-v2/dockercommandutils";
+import * as fileUtils from "docker-common-v2/fileutils";
+import * as utils from "./utils";
+
+export async function runBuild(connection: ContainerConnection) {
+    // find dockerfile path
+    let dockerfilepath = tl.getInput("Dockerfile", true);
+    let dockerFile = fileUtils.findDockerFile(dockerfilepath);
+    
+    if(!tl.exist(dockerFile)) {
+        throw new Error(tl.loc('ContainerDockerFileNotFound', dockerfilepath));
+    }
+  
+    // get qualified image names by combining container registry(s) and repository
+    let repositoryName = tl.getInput("repository");
+    let imageNames: string[] = [];    
+    // if container registry is provided, use that
+    // else, use the currently logged in registries
+    if (tl.getInput("dockerRegistryServiceConnection")) {
+        let imageName = connection.getQualifiedImageName(repositoryName, true);
+        if (imageName) {
+            imageNames.push(imageName);
+        }
+    }
+    else {
+        imageNames = connection.getQualifiedImageNamesFromConfig(repositoryName, true);
+    }
+
+    // get tags input
+    let tags = tl.getDelimitedInput("tags", "\n");
+    let tagArguments: string[] = [];
+    // find all the tag arguments to be added to the command
+    if (imageNames && imageNames.length > 0) {
+        imageNames.forEach(imageName => {
+            if (tags && tags.length > 0) {
+                tags.forEach(tag => {
+                    tagArguments.push(imageName + ":" + tag);
+                });
+            }
+            else {
```

Comment:
As I had specified in the spec, if the user has explicitly provided the tags input, we honour that. Else, the default value of $(Build.BuildId) needs to be used

Reply:
@shashankbarsin, This current behavior is same as other tasks. we should not default to $(Build.BuildId) if user gives empty string. If we always generates default tag, we will get issues from customer.

Reply:
Today if user provides the following task input, $(Build.BuildId) is used for tagging the image because the task input has a default value of $(Build.BuildId) in task.json. And if the input is not provided a value, this default of $(Build.BuildId) is used. 

```YAML
- task: Docker@2
    inputs:
      containerRegistry: 'shasb'
      repository: 'shashankbarsin/playground'
      command: 'buildAndPush'
      Dockerfile: '**/Dockerfile'
```

In this context, which case is this else clause executed for? Only for the case where user explicitly mentions a tag input, but provides an empty string as value? Like so - 

```YAML
- task: Docker@2
    inputs:
      containerRegistry: 'shasb'
      repository: 'shashankbarsin/playground'
      command: 'buildAndPush'
      Dockerfile: '**/Dockerfile'
      tags: 
```

If that is the case, it's better to have consistency between the two cases I've mentioned above by using $(Build.BuildId) in both cases at least for this task as there's no issue of backward compatibility break

Reply:
> Today if user provides the following task input, $(Build.BuildId) is used for tagging the image because the task input has a default value of $(Build.BuildId) in task.json. And if the input is not provided a value, this default of $(Build.BuildId) is used.
> - task: Docker@2
>     inputs:
>       containerRegistry: 'shasb'
>       repository: 'shashankbarsin/playground'
>       command: 'buildAndPush'
>       Dockerfile: '**/Dockerfile'
> In this context, which case is this else clause executed for?  

Here input it self not specified so server sending the default input  so it will tag with that. 

Only for the case where user explicitly mentions a tag input, but provides an empty string as value? Like so -
> - task: Docker@2
>     inputs:
>       containerRegistry: 'shasb'
>       repository: 'shashankbarsin/playground'
>       command: 'buildAndPush'
>       Dockerfile: '**/Dockerfile'
>       tags: 
> If that is the case, it's better to have consistency between the two cases I've mentioned above by using $(Build.BuildId) in both cases at least for this task as there's no issue of backward compatibility break

In this case user specified tags with empty input. I feel we should not tag anything.  Adding $(Build.BuildId) is easy if customers reports they don't that behavior them removing is difficult. So I feel it's not better to tag if input is empty.